server/deviceflowhandlers.go

```diff
@@ -71,6 +71,12 @@ func (s *Server) handleDeviceCode(w http.ResponseWriter, r *http.Request) {
 		clientID := r.Form.Get("client_id")
 		clientSecret := r.Form.Get("client_secret")
 		scopes := strings.Fields(r.Form.Get("scope"))
+		codeChallenge := r.Form.Get("code_challenge")
+		codeChallengeMethod := r.Form.Get("code_challenge_method")
+
+		if codeChallengeMethod == "" {
```

Comment:
I've found a problem here. The cause is that we do not validate `code_challenge_method` on parsing a device auth request.

How to reproduce:
1. Request the `/device/code` endpoint.There will be no error in the response.
```sh
 curl --request POST \
  --url http://127.0.0.1:5556/dex/device/code \
  --header 'Accept: application/json' \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --data-urlencode 'client_id=example-device-client' \
  --data-urlencode 'scope=openid profile offline_access' \
  --data-urlencode 'code_challenge=test' \                                      
  --data-urlencode 'code_challenge_method=ivalid-method' 
```
2. Authenticate.
3. Get the token.
```sh
curl --request POST \                              
  --url http://127.0.0.1:5556/dex/token \
  --header 'Accept: application/json' \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --data-urlencode 'client_id=example-device-client' \
  --data-urlencode 'scope=openid profile offline_access' \
  --data-urlencode 'grant_type=urn:ietf:params:oauth:grant-type:device_code' \
  --data-urlencode 'device_code=xxxxxxxxxxxxxxxxxxx' \
  --data-urlencode 'code_verifier=test'
```
4. There will be a `500 Internal Server Error` in response.
```
< HTTP/1.1 500 Internal Server Error
< Content-Length: 24
< Content-Type: application/json
< Date: Tue, 26 Jul 2022 14:05:04 GMT
< 
* Connection #0 to host 127.0.0.1 left intact
{"error":"server_error"}%  
```

I think we must validate the `code_challenge_method` values is either `plain`, `S256`, or empty. Corresponding validation for the authcode grant can be found here:
https://github.com/dexidp/dex/blob/50dc2f55187a0e72d73809485725e45eba42b4eb/server/oauth2.go#L475-L478

Reply:
It is validated on the token exchange side, but I added a check on the device code endpoint to fail earlier if the value is invalid.